src/middleware/builtin.ts

```diff
@@ -79,100 +78,73 @@ export function onlyEvents(): Middleware<AnyMiddlewareArgs & { event?: SlackEven
   };
 }
 
-/**
- * Helper function that determines if a payload is an options payload
- */
-function isOptionsPayload(payload: KeyValueMapping): boolean {
-  if (payload.type !== 'dialog_suggestion' && !isInteractiveMessageOptions(payload) && !isExternalSelect(payload)) {
-    return false;
-  }
-  return true;
-}
-
-/**
- * Typeguard that decides if a payload is an interactive message options request
- */
-function isInteractiveMessageOptions(payload: KeyValueMapping):
-payload is ExternalOptionsRequest<'interactive_message'> {
-  if ((payload as InteractiveMessage<InteractiveAction>).actions) {
-    return false;
-  }
-  if (payload.type !== 'interactive_message') {
-    return false;
-  }
-  return true;
-}
-
-/**
- * Typeguard that decides if a payload is an interactive message options request
- */
-function isExternalSelect(payload: KeyValueMapping):
-payload is Actions<ExternalSelectResponse> {
-  if ((payload as Actions<ExternalSelectResponse>).actions &&
-    (payload as Actions<ExternalSelectResponse>).actions[0].type === 'external_select') {
-    return true;
-  }
-  return false;
-}
-
-/**
- * Typeguard that decides whether a constraint object is an options constraint object
- */
-function hasCallbackId(payload: { [key: string]: any; }): payload is (
-    InteractiveMessage<InteractiveAction> | DialogSubmitAction | MessageAction | ExternalOptionsRequest<Container>
-  ) {
-  if ((payload as (InteractiveMessage<InteractiveAction> | DialogSubmitAction |
-    MessageAction | ExternalOptionsRequest<Container>)).callback_id) {
-    return true;
-  }
-  return false;
-}
-
 /**
  * Middleware that checks for matches given constraints
  */
 export function matchActionConstraints(
     constraints: ActionConstraints,
+    // TODO: this function signature could be wrong. this gets used in options() so the action property can be undefined
   ): Middleware<SlackActionMiddlewareArgs<SlackAction>> {
   return ({ action, next, context }) => {
+    // NOTE: DialogSubmitAction and MessageAction do not have an actions array
+    const innerAction: InteractiveAction | UnknownElementAction<string> | undefined =
+      action.actions !== undefined ? action.actions[0] : undefined;
+
+    // TODO: is putting matches in an array actually helpful? there's no way to know which of the regexps contributed
```

Comment:
https://github.com/slackapi/node-slack-interactive-messages/issues/47

Reply:
This is why I used this. If we don't want them, we can switch to using `test()` instead and get rid of matches array. I could see this particularly helpful when pulling out data like userIDs or list of fixed-structure data (like JIRA tasks would all be JIRA-xxx).

Reply:
i think matches is a great idea! and i think we even had this in the gist when talking about the text pattern matching for message events. i'm specifically questioning if `context.matches` should be an `any[]`, or if it should have some inner structure. the problem with a flat array is that if i use both a RegExp for `block_id` and a RegExp for `action_id`, how do i know which of the matches are for which? even if i know `block_id` matches come first, how do I know at which index those end and the `action_id` matches begin?